### ISSUE TITLE
[LifetimeSafety] False positive return stack address and dangling assignment with custom pointer class

### ISSUE THREAD

**Issue by AMDG2** [llvm/llvm-project#188481]
The following code produces two false positive warnings. I've included similar functions that uses `std::string_view` instead of the custom `gsl::Pointer` type to show that the false positives are not present with it.

https://godbolt.org/z/57PM46WT7

```c++
#include <memory>

template<typename T>
struct [[gsl::Owner(T)]] Optional;

template <typename T>
struct [[gsl::Pointer(T)]] Optional<T&> {

    T* ptr = nullptr;

    Optional() = default;
    Optional(T& value) : ptr(std::addressof(value)) {}
    
    // Enabling this workaround the false positive -Wdangling-assignment-gsl on line 31.
#if 0
    Optional& operator=(T& value) {
        ptr = std::addressof(value);
        return *this;
    }
#endif
};

Optional<const int&> foo1() {
    Optional<const int&> empty;
    return empty; // false positive -Wreturn-stack-address
}

Optional<const int&> foo2() {
    Optional<const int&> stackPtr;
    int foo = 10;
    stackPtr = foo; // false positive -Wdangling-assignment-gsl
    return stackPtr; // correct -Wreturn-stack-address
}

std::string_view sv1() {
    std::string_view empty;
    return empty;
}

std::string_view sv2() {
    std::string_view stackStr;
    std::string str = "Hello";
    stackStr = str; // correct -Wlifetime-safety-return-stack-addr
    return stackStr;
}
```

**Comment by AMDG2** [comment]
I've updated to Clang 23.1.0 and I can confirm the issues are solved as expected.

The diagnostics `-Wdangling-assignment-gsl` and `-Wreturn-stack-address` still have the false positives I reported. However, this is a **won't fix** from your side, which is fair as  `-Wlifetime-safety` fully covers those cases in my testing, and the error messages are even improved.
